cmd/job_run_list.go

```diff
@@ -0,0 +1,156 @@
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/odpf/salt/log"
+	cli "github.com/spf13/cobra"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	pb "github.com/odpf/optimus/api/proto/odpf/optimus/core/v1beta1"
+)
+
+const (
+	jobStatusTimeout = time.Second * 30
+)
+
+type jobRunCmdArg struct {
+	projectName string
+	jobName     string
+	startDate   string
+	endDate     string
+}
+
+type dateRange struct {
+	start time.Time
+	end   time.Time
+}
+
+type getRequestFn func(jobRunCmdArg) (*pb.JobRunRequest, error)
+
+func jobRunListCommand(l log.Logger, defaultProjectName, defaultHost string) *cli.Command {
+	cmd := &cli.Command{
+		Use:     "runs",
+		Short:   "Get current job runs",
+		Example: `optimus job runs <sample_job_goes_here> [--project \"project-id\"] [--start_date \"2006-01-02T15:04:05Z07:00\" --end_date \"2006-01-02T15:04:05Z07:00\"]`,
+		Args:    cli.MinimumNArgs(1),
+	}
+	projectName := defaultProjectName
+	host := defaultHost
+	var startDate string
+	var endDate string
+	cmd.Flags().StringVarP(&projectName, "project", "p", projectName, "Project name of optimus managed repository")
+	cmd.Flags().StringVar(&host, "host", defaultHost, "Optimus service endpoint url")
+	cmd.Flags().StringVar(&startDate, "start_date", "", "start date of job run")
+	cmd.Flags().StringVar(&endDate, "end_date", "", "end date of job run")
+	cmd.RunE = func(c *cli.Command, args []string) error {
+		jobName := args[0]
+		l.Info(fmt.Sprintf("Requesting status for project %s, job %s from %s",
+			projectName, jobName, host))
+		arg := jobRunCmdArg{
+			projectName: projectName,
+			jobName:     jobName,
+			startDate:   startDate,
+			endDate:     endDate,
+		}
+		return getJobRunList(l, host, arg)
+	}
+	return cmd
+}
+
+func getJobRunList(l log.Logger, host string, arg jobRunCmdArg) error {
+	// get the last schedule run
+	if arg.startDate == "" && arg.endDate == "" {
+		return callJobRun(l, host, arg, getJobRunRequest)
+	}
+	// get runs between start_date and end_date (both start_date and end_date are inclusive)
+	return callJobRun(l, host, arg, getJobRunRequestWithDate)
+}
+
+func callJobRun(l log.Logger, host string, arg jobRunCmdArg, fn getRequestFn) error {
+	var err error
+	dialTimeoutCtx, dialCancel := context.WithTimeout(context.Background(), OptimusDialTimeout)
```

Comment:
isn't it a best practice to create the context in the command and pass it to the needed functions from there instead of creating the context in the sub functions.

Reply:
context is passed from `getJobRunList` now